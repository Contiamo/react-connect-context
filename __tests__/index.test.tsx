import * as React from "react"
import renderer from "react-test-renderer"
import { connectContext, MapContextToProps, MergeProps } from "../"
import "jest"

interface ContentProps {
  color: string
  myProp: string
}

interface ContextValue {
  color: string
}

const Context = React.createContext<ContextValue>({ color: "red" })
const Content: React.SFC<ContentProps> = ({ children, color, myProp }) => (
  <div>
    <p>{children}</p>
    <div>
      I have looked into context and I've seen the color is: {color}
      I also get my own props like {myProp}!
    </div>
  </div>
)

describe("connectContext", () => {
  it("should map context to props while preserving existing props", () => {
    const ConnectedContent = connectContext<ContextValue, ContentProps>(
      Context.Consumer
    )(Content)

    const App = () => (
      <div>
        <ConnectedContent myProp="test">Hello</ConnectedContent>
      </div>
    )

    const tree = renderer
      .create(
        <Context.Provider value={{ color: "green" }}>
          <App />
        </Context.Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("Should use the custom mapContextToProps function if provided", () => {
    const mapContextToProps: MapContextToProps<
      ContextValue,
      ContentProps
    > = ({ color }) => {
      return ({ color: color === "green" ? "#00ff00" : color })
    }

    const ConnectedContent = connectContext(
      Context.Consumer,
      mapContextToProps
    )(Content)

    const App = () => (
      <ConnectedContent myProp="sup">
        Sup my frand
      </ConnectedContent>
    )

    const tree = renderer
      .create(
        <Context.Provider value={{ color: "green" }}>
          <App />
        </Context.Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("Should use the custom mergeProps function if provided", () => {
    const mergeProps: MergeProps<
      ContextValue,
      ContentProps
    > = (context, props) => {
      return ({ ...props, ...context })
    }

    const ConnectedContent = connectContext(
      Context.Consumer,
      undefined,
      mergeProps
    )(Content)

    const AppWithPropsConflict = () => (
      <ConnectedContent myProp="sup" color="red">
        Sup my frand
      </ConnectedContent>
    )

    const tree = renderer
      .create(
        <Context.Provider value={{ color: "green" }}>
          <AppWithPropsConflict />
        </Context.Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
