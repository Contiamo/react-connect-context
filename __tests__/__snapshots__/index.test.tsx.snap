// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectContext Should use the custom mapContextToProps function if provided 1`] = `
<div>
  <p>
    Sup my frand
  </p>
  <div>
    I have looked into context and I've seen the color is: 
    #00ff00
    I also get my own props like 
    sup
    !
  </div>
</div>
`;

exports[`connectContext Should use the custom mergeProps function if provided 1`] = `
<div>
  <p>
    Sup my frand
  </p>
  <div>
    I have looked into context and I've seen the color is: 
    green
    I also get my own props like 
    sup
    !
  </div>
</div>
`;

exports[`connectContext should map context to props while preserving existing props 1`] = `
<div>
  <div>
    <p>
      Hello
    </p>
    <div>
      I have looked into context and I've seen the color is: 
      green
      I also get my own props like 
      test
      !
    </div>
  </div>
</div>
`;
